### DISCUSSION TITLE
Lyra 4.2 I2C Pins

### DISCUSSION THREAD

**Discussion by jonrichings** [pschatzmann/arduino-audiokit#21] [General]
Most of AudioKit seems to be OK so I assumed that anything I2C was using the correct pins.  However I just tried an external I2C device and it does not work with the standard Lyra/AudioKit setup.  

It does work just running it's example code on the Lyra, if I use #including wire.h and Wire.begin(18, 23);   Adding that to an AudioKit example doesn't work.  #including wire.h and Wire.begin(18, 23); crashes.  Must be conflicting with AudioKit's 12C config.

Am I doing something wrong?  Can I change the pins in a .h somewhere, or can I override it in the .ino?

**Comment by pschatzmann** [comment]
I suggest that you __update to the latest version__ and then read https://github.com/pschatzmann/arduino-audiokit/wiki/SPI-and-I2C.

If you have an old version, channel 0 (which corresponds to the Wire object) is used by the AudioKit and must not be used.

**Reply by jonrichings** [reply]
OK, I updated to the latest release.  But I think there is still a problem here.  The audio chip uses pins 18 and 23 and these are also brought out to the I2C pin header.  Pins 21 and 22 (which ESP32 I2C would normally use) are wired to a led and the PA enable.  So I2C on 18 and 23 is the only I2C physically available.  Any additional I2C device still has to be able to use those pins.  Can Wire and Wire1 both use 18 and 23?.  Or am I misunderstanding your description on https://github.com/pschatzmann/arduino-audiokit/wiki/SPI-and-I2C ?

**Comment by pschatzmann** [comment]
I think you might be able to use Wire with pin 13 and 15 on the jtag section, if you don't use the SD.
Since Wire1 is already set up to use 18 and 23  you should be able to use this object w/o any additional initialisation to do your communication. If you use a library you might need to select one which accepts a TwoWire object as agrument...

**Reply by jonrichings** [reply]
If I understand you correctly, you use Wire1 EXCLSUIVELY for the audio chip?  It can no longer be used as a bus for other I2C devices?

In my application I could maybe use the default pins 20 and 21 if remove them from the LED and PA enable.  Would that work?

**Comment by pschatzmann** [comment]
You should be able to reuse Wire1. Just make sure that you don't call any additoinal begin on it...

**Reply by jonrichings** [reply]
And then modify the libs to use Wire1?

**Reply by jonrichings** [reply]
FYI, calling Wire.begin(); crashes.

**Reply by pschatzmann** [reply]
Did you indicate the pins when you call begin? 
If so which ones?
What exactly are you trying to do in the end ?

**Reply by jonrichings** [reply]
Good catch.  I did not specify the pins.  Wire.begin(21, 22); does not crash.  But now I have to get my strongest glasses out and hard wire those pins and disconnect the PA and LED.  I'll let you know how that goes.

In the end I am building an electronic flute that uses touch sensors for the "keys" and a pressure sensor for the breath control.  I may also add an accelerometer for extra performance control and a haptic device for some "realistic" feedback.  All are likely to be I2C devices.  All wrangled by your code and sound generated by Faust.

BTW, why did you put the audio chip in Wire1?

**Reply by jonrichings** [reply]
OK, I am being dumb.  I could hardwire to the 21 and 22 pins and I could live without the PA's and the green LED.  But your HAL would still be expecting to drive those, right?  And, depending on what starts up first, those things may prevent Wire from driving those pins.  So I guess my only option is to rework all the peripheral drivers to work with Wire1?

**Comment by pschatzmann** [comment]
I don't do anything with the green LED, but I manage the PA to activate the speaker and I am not sure if it is a good idea to use this pin.

I admit, I did not realize how bad the situation with the available GPIOs really is. I would consider the following pins:
- IO03 RX0  
- IO14? MTMS 
- IO19 KEY3 
- IO22 GREEN_LED? 

For the other option I would recommend to prefer drivers that don't do any I2C initialization and allow you to pass a TwoWire object.

**Reply by jonrichings** [reply]
I tried the JTAG pins with no luck (maybe there is data flowing there already) and IO33 and IO34 (since I don't plan to use the onboard touch sensors).  The simple test example for the device works fine with these pins.  And in my/your code Setup() does see the I2C devices but as soon as it gets to the main loop it crashes.

Here's my code.  I commented out the BLE to be sure it wasn't that.
```

/**
 * @file streams-faust-i2s.ino
 * @author Phil Schatzmann
 * @brief Example how to use Faust as "Singal Processor"
 * @version 0.1
 * @date 2022-04-22
 * 
 * @copyright Copyright (c) 2022
 * 
 */

#include "AudioTools.h"
#include "AudioLibs/AudioKit.h"
#include "AudioLibs/AudioFaust.h"
#include "volume.h"
//#include <BLEMidi.h>
//#include <Wire.h>

#include "Adafruit_MPR121.h"

#ifndef _BV
#define _BV(bit) (1 << (bit)) 
#endif

// You can have up to 4 on one i2c bus but one is enough for testing!
Adafruit_MPR121 cap = Adafruit_MPR121();

// Keeps track of the last pins touched
// so we know when buttons are 'released'
uint16_t lasttouched = 0;
uint16_t currtouched = 0;


// streams setup
SineWaveGenerator<int16_t> sineWave(32000);                // subclass of SoundGenerator with max amplitude of 32000
GeneratedSoundStream<int16_t> sound(sineWave);             // Stream generated from sine wave
mydsp dsp;
AudioKitStream out; 
FaustStream faust(dsp, out);
StreamCopy copier(out, sound);  //

// Arduino Setup
void setup(void) {  
  
  Wire.begin(33, 32);  // note this is different from the default usually shown for esp32.  I think this is a Lyra specific thing.
  // Open Serial 
  Serial.begin(115200);
//  AudioLogger::instance().begin(Serial, AudioLogger::Warning);

  while (!Serial) { // needed to keep leonardo/micro from starting too fast!
    delay(10);
  }
  
  Serial.println("Adafruit MPR121 Capacitive Touch sensor test"); 
  
  // Default address is 0x5A, if tied to 3.3V its 0x5B
  // If tied to SDA its 0x5C and if SCL then 0x5D
  if (!cap.begin(0x5A)) {
    Serial.println("MPR121 not found, check wiring?");
    while (1);
  }
  Serial.println("MPR121 found!");

  
  // Setup Faust volume control
  auto cfg = faust.defaultConfig();
  faust.begin(cfg);
  faust.setLabelValue("0x00", -20);

  // Setup sine wave generator
  sineWave.begin(cfg.channels, cfg.sample_rate, N_B4);
  //sineWave.setFrequency(0);

  // setup ble-midi stuff // 
  /*
  BLEMidiServer.begin("My MIDI Device");
  BLEMidiServer.setOnConnectCallback(onConnect);
  BLEMidiServer.setOnDisconnectCallback(onDisconnect);
  BLEMidiServer.setNoteOnCallback(onNoteOn);
  BLEMidiServer.setNoteOffCallback(onNoteOff);
  BLEMidiServer.setControlChangeCallback(onControlChange);
  BLEMidiServer.setAfterTouchPolyCallback(onAfterTouchPoly);
  BLEMidiServer.setProgramChangeCallback(onProgramChange);
  BLEMidiServer.setAfterTouchCallback(onAfterTouch);
  BLEMidiServer.setPitchBendCallback(onPitchbend);
  BLEMidiServer.enableDebugging();
  */

  // start I2S
  auto cfg_i2s = out.defaultConfig(TX_MODE);
  cfg_i2s.sample_rate = cfg.sample_rate; 
  cfg_i2s.channels = cfg.channels;
  cfg_i2s.bits_per_sample = cfg.bits_per_sample;
  out.begin(cfg_i2s);

}

// Arduino loop - copy sound to out 
void loop() {
  copier.copy();

/*
  // ble-midi client ticker
//    if (BLEMidiServer.isConnected()) {
      //BLEMidiServer.noteOn(0, 69, 127);
      //Serial.printf("Note on");
      //delay(1000);
      //BLEMidiServer.noteOff(0, 69, 127);
      //Serial.printf("Note on");
      //delay(1000);
  }


    currtouched = cap.touched();
  
  for (uint8_t i=0; i<12; i++) {
    // it if *is* touched and *wasnt* touched before, alert!
    if ((currtouched & _BV(i)) && !(lasttouched & _BV(i)) ) {
      Serial.print(i); Serial.println(" touched");
    }
    // if it *was* touched and now *isnt*, alert!
    if (!(currtouched & _BV(i)) && (lasttouched & _BV(i)) ) {
      Serial.print(i); Serial.println(" released");
    }
  }

  // reset our state
  lasttouched = currtouched;

  // comment out this line for detailed data from the sensor!
  return;
  
  // debugging info, what
  Serial.print("\t\t\t\t\t\t\t\t\t\t\t\t\t 0x"); Serial.println(cap.touched(), HEX);
  Serial.print("Filt: ");
  for (uint8_t i=0; i<12; i++) {
    Serial.print(cap.filteredData(i)); Serial.print("\t");
  }
  Serial.println();
  Serial.print("Base: ");
  for (uint8_t i=0; i<12; i++) {
    Serial.print(cap.baselineData(i)); Serial.print("\t");
  }
  Serial.println();
  
  // put a delay so it isn't overwhelming
  delay(100);

*/
  
}


// ble-midi callbacks
void onConnect()
{
  Serial.println("Connected");
}

void onDisconnect()
{
  Serial.println("Disconnected");
}

void onNoteOn(uint8_t channel, uint8_t note, uint8_t velocity, uint16_t timestamp)
{
  float midiNote = note;
  
  //int16_t freq = MidiCommon::noteToFrequency(note);
  int16_t freq  =  440 * pow(2, (midiNote-69)/12);
  Serial.printf("Note on : channel %d, note %d, velocity %d, freq %d, (timestamp %dms)\n", channel, note, velocity, freq, timestamp);
  sineWave.setFrequency(freq);
}

void onNoteOff(uint8_t channel, uint8_t note, uint8_t velocity, uint16_t timestamp)
{
  Serial.printf("Note off : channel %d, note %d, velocity %d (timestamp %dms)\n", channel, note, velocity, timestamp);
  sineWave.setFrequency(0);
}

void onAfterTouchPoly(uint8_t channel, uint8_t note, uint8_t pressure, uint16_t timestamp)
{
  Serial.printf("Polyphonic after touch : channel %d, note %d, pressure %d (timestamp %dms)\n", channel, note, pressure, timestamp);
}

void onControlChange(uint8_t channel, uint8_t controller, uint8_t value, uint16_t timestamp)
{
    Serial.printf("Control change : channel %d, controller %d, value %d (timestamp %dms)\n", channel, controller, value, timestamp);
}

void onProgramChange(uint8_t channel, uint8_t program, uint16_t timestamp)
{
    Serial.printf("Program change : channel %d, program %d (timestamp %dms)\n", channel, program, timestamp);
}

void onAfterTouch(uint8_t channel, uint8_t pressure, uint16_t timestamp)
{
    Serial.printf("After touch : channel %d, pressure %d (timestamp %dms)\n", channel, pressure, timestamp);
}

void onPitchbend(uint8_t channel, uint16_t value, uint16_t timestamp)
{
    Serial.printf("Pitch bend : channel %d, value %d (timestamp %dms)\n", channel, value, timestamp);
}
```

**Reply by jonrichings** [reply]
Some better clues:

I made a small change and that triggered a different crash message:

```
rst:0x1 (POWERON_RESET),boot:0x1f (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:1
load:0x3fff0030,len:1324
ho 0 tail 12 room 4
load:0x40078000,len:13508
load:0x40080400,len:3604
entry 0x400805f0
Starting up sensors
MPR121 found!
Starting up Faust
Starting up sinewave
Starting up BLE
Starting I2S
[W] AudioKit.h : 43 - sd_active = true
E (939) i2c: i2c driver install error
Error:   AC101
Error:   reset failed!
Error:   codec init failed!
Error:   audio_hal_init
[E] AudioKit.h : 185 - begin faild: please verify your AUDIOKIT_BOARD setting: 6
```

For reference I have my I2C stuff setup as Wire.begin(33, 32); which works with non audiokit stuff.

I'm thinking that the "AC101" reference in the error is the problem.  Maybe audiokit is expecting an AC101 audio chip when Lyra uses ES8388?  So it's getting no response?

**Reply by jonrichings** [reply]
Duh.  I almost answered my own problem.  Yes AC101 is the problem.  For future reference the AudioKitSettings.h needs to be edited to use the correct board ID.  Line 22 needed to be changed from 6 to 1.  I hadn't edited it after updating the files.

FYI, if someone reading this is an Arduino fan you may be used to not even looking at all the files that Arduino code includes (and the includes that they include), but occasionally it is necessary.  This is one of them.

**Reply by lijianyu1985** [reply]
@jonrichings did you fix this error?

**Reply by pschatzmann** [reply]
This is not an error: I suggest that you read the Readme of the project and maybe the FAQ
